src/Storages/MergeTree/MergeTreeData.cpp

```diff
@@ -4951,6 +4949,18 @@ MergeTreeData::DataPartsVector MergeTreeData::Transaction::commit(MergeTreeData:
         auto parts_lock = acquired_parts_lock ? MergeTreeData::DataPartsLock() : data.lockParts();
         auto * owing_parts_lock = acquired_parts_lock ? acquired_parts_lock : &parts_lock;
 
+        if (txn)
+        {
+            for (const DataPartPtr & part : precommitted_parts)
+            {
+                DataPartPtr covering_part;
+                DataPartsVector covered_parts = data.getActivePartsToReplace(part->info, part->name, covering_part, *owing_parts_lock);
+                MergeTreeTransaction::addNewPartAndRemoveCovered(data.shared_from_this(), part, covered_parts, txn);
+            }
+        }
+
+        NOEXCEPT_SCOPE;
```

Comment:
Can you please explain why it is required? And looks like it is better to add a comment for users of NOEXCEPT_SCOPE, thoughts?

Reply:
See the comment:
https://github.com/ClickHouse/ClickHouse/blob/39c0219c11aab1d4fd0897d05e00543eeef45eef/src/Common/noexcept_scope.h#L30-L33

It's required in `MergeTreeData::Transaction::commit` because changes to precommitted parts and `MergeTreeData` state (like set of parts with their states, indexes and column sizes) are not undone on exceptions. Normally this code should not throw any exceptions (unless maybe `MEMORY_LIMIT_EXCEEDED`, but exceptions from `MemoryTracker` are blocked), but in case it does (due to some bug or whatever) I would prefer to terminate immediately. 

You can think of this macro as kind of paranoid defensive programming :)